### PR TITLE
Kafka upgrade (#2871) (#2803)

### PR DESCRIPTION
**Pull request by cicharka**
* Kafka: Upgrade to 2.8.1 version

* Kafka: Refactor installation and upgrade tasks

* Kafka: move upgrade tasks to Kafka role

* Zookeeper: refactor installation and upgrade

* Zookeeper: move upgrade tasks to Zookeeper role

* Kafka: Refactor: rename 'kafka_var' setting

* Kafka-exporter: verify if service definition needs to be updated even if exporter itself was not updated

* Kafka-exporter: remove kafka.version from configuration

### CODE REVIEW

**Comment by przemyslavic** [comment]
Don't we want to update ZooKeeper to v3.5.9? https://zookeeper.apache.org/doc/r3.5.9/releasenotes.html
It comes with Kafka 2.8.0.

**Comment by cicharka** [comment]
> Don't we want to update ZooKeeper to v3.5.9? https://zookeeper.apache.org/doc/r3.5.9/releasenotes.html It comes with Kafka 2.8.0.

I think that I didn't take it into account because:
* documentation of Ecosystem for Kafka 2.8.1 says nothing about Zookeeper upgrade -> it only mention that in Kafka 2.5.0 ZooKeeper has been upgraded to 3.5.7. Additionally I have not checked Kafka archive in order to investigate Zookeeper version which comes with it (didn't know about it to be honest)
* I checked that Kafka doesn't specify Zookeeper version required, and as we are installing it as standalone component and I have not found any issues that states to upgrade it, I just skipped it

However, since you asked about it, and even found that it comes with Kafka, we can think about 2 options:
1. We are going to remove ZooKeeper role since it is only used as prerequisite for Kafka (for my understanding, if it's not true please correct me) and we are going to use Zookeeper that comes with Kafka (additionally Kafka states that they already removed ZooKeeper from Kafka, but still in development stage)
2. We are going only to upgrade ZooKeeper version -> ZooKeeper role will stay, only defaults and binary changed needed.

I don't know about any other ZooKeeper usecases but if its only Kafka's prerequisite I would vote for role removal and move to use ZooKeeper that comes with Kafka. (I would also do it in another PR to do not overwhelm reviewers :) )
Sorry for long post!

File: ansible/playbooks/roles/kafka/tasks/common/download_and_unpack_binary.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+---
+
```
Comment (rafzei) [inline ansible/playbooks/roles/kafka/tasks/common/download_and_unpack_binary.yml]:
```suggestion

```

File: ansible/playbooks/roles/kafka/tasks/common/stop.yml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+
```
Comment (rafzei) [inline ansible/playbooks/roles/kafka/tasks/common/stop.yml]:
```suggestion

```

File: ansible/playbooks/roles/kafka/tasks/generate-certificates.yml [hunk 3]
```diff
@@ -1,55 +1,55 @@
 - name: Create stores directory
   file:
-    path: "{{ specification.kafka_var.security.ssl.server.keystore_location | dirname }}"
+    path: "{{ specification.security.ssl.server.keystore_location | dirname }}"
     state: directory
-    owner: "{{ specification.kafka_var.user }}"
-    group: "{{ specification.kafka_var.group }}"
+    owner: "{{ specification.user }}"
+    group: "{{ specification.group }}"
     mode: "0755"
```
Comment (rafzei) [inline ansible/playbooks/roles/kafka/tasks/generate-certificates.yml]: Some time ago we agreed to use symbolic modes
```suggestion
    mode: u=rwx,go=rx
```
BTW do we need that wide permissions? (please check also in other places)

File: ansible/playbooks/roles/kafka/tasks/generate-certificates.yml [hunk 4]
```diff
@@ -1,55 +1,55 @@
...
+    path: "{{ specification.security.ssl.server.keystore_location }}"
   changed_when: false
   register: keystore_exists
 
 - name: Generate keystore for each server
-  shell: keytool -keystore {{ specification.kafka_var.security.ssl.server.keystore_location }} \
-         -alias localhost -validity {{ specification.kafka_var.security.ssl.server.cert_validity }} -genkey -keyalg RSA \
-         -noprompt -storepass {{ specification.kafka_var.security.ssl.server.passwords.keystore }} \
-         -keypass {{ specification.kafka_var.security.ssl.server.passwords.key }} \
+  shell: keytool -keystore {{ specification.security.ssl.server.keystore_location }} \
```
Comment (rafzei) [inline ansible/playbooks/roles/kafka/tasks/generate-certificates.yml]: I would use `command` here and in other places w/o chars like <,>, |  and &

File: ansible/playbooks/roles/kafka/tasks/metrics.yml [hunk 5]
```diff
@@ -2,17 +2,17 @@
...
-  become: yes
+  become: true
   copy:
     dest: "{{ specification.prometheus_jmx_config }}"
     src: jmx-kafka-config.yml
-    owner: "{{ specification.kafka_var.user }}"
-    group: "{{ specification.kafka_var.group }}"
+    owner: "{{ specification.user }}"
+    group: "{{ specification.group }}"
     mode: 0644
```
Comment (rafzei) [inline ansible/playbooks/roles/kafka/tasks/metrics.yml]:
```suggestion
    mode: u=rx,go=r
```

File: ansible/playbooks/roles/kafka/tasks/setup-kafka.yml [hunk 6]
```diff
@@ -24,48 +24,20 @@
...
-  vars:
-    file_name: "{{ kafka_file_name }}"
+- name: Download and unpack Kafka's binary
+  include_tasks: common/download_and_unpack_binary.yml
 
 - name: Add Kafka's bin dir to the PATH
   copy:
     content: "export PATH=$PATH:/opt/kafka/bin"
     dest: "/etc/profile.d/kafka_path.sh"
     mode: 0755
```
Comment (rafzei) [inline ansible/playbooks/roles/kafka/tasks/setup-kafka.yml]:
```suggestion
    mode: u=rwx,go=rx
```

File: ansible/playbooks/roles/kafka/tasks/setup-kafka.yml [hunk 7]
```diff
@@ -78,59 +50,60 @@
...
 - name: Create data_dir
   file:
-    path: "{{ specification.kafka_var.data_dir }}"
+    path: "{{ specification.data_dir }}"
     state: directory
-    owner: "{{ specification.kafka_var.user }}"
-    group: "{{ specification.kafka_var.group }}"
+    owner: "{{ specification.user }}"
+    group: "{{ specification.group }}"
     mode: 0755
```
Comment (rafzei) [inline ansible/playbooks/roles/kafka/tasks/setup-kafka.yml]:
```suggestion
    mode: u=rwx,go=rx
```

File: ansible/playbooks/roles/kafka/tasks/setup-kafka.yml [hunk 8]
```diff
@@ -78,59 +50,60 @@
...
 - name: Create log_dir
   file:
-    path: "{{ specification.kafka_var.log_dir }}"
+    path: "{{ specification.log_dir }}"
     state: directory
-    owner: "{{ specification.kafka_var.user }}"
-    group: "{{ specification.kafka_var.group }}"
+    owner: "{{ specification.user }}"
+    group: "{{ specification.group }}"
     mode: 0755
```
Comment (rafzei) [inline ansible/playbooks/roles/kafka/tasks/setup-kafka.yml]:
```suggestion
    mode: u=rwx,go=rx
```

File: ansible/playbooks/roles/kafka/tasks/setup-kafka.yml [hunk 9]
```diff
@@ -78,59 +50,60 @@
...
 # Setup log4j.properties
 - name: Create log4j.properties
   file:
-    dest: "{{ specification.kafka_var.conf_dir }}/log4j.properties"
-    owner: "{{ specification.kafka_var.user }}"
-    group: "{{ specification.kafka_var.group }}"
+    path: "{{ specification.conf_dir }}/log4j.properties"
+    owner: "{{ specification.user }}"
+    group: "{{ specification.group }}"
     mode: 0644
```
Comment (rafzei) [inline ansible/playbooks/roles/kafka/tasks/setup-kafka.yml]:
```suggestion
    mode: u=rw,go=r
```

File: ansible/playbooks/roles/kafka/tasks/setup-kafka.yml [hunk 10]
```diff
@@ -78,59 +50,60 @@
...
 - name: Create server.properties
   template:
-    dest: "{{ specification.kafka_var.conf_dir }}/server.properties"
-    owner: "{{ specification.kafka_var.user }}"
-    group: "{{ specification.kafka_var.group }}"
+    dest: "{{ specification.conf_dir }}/server.properties"
+    owner: "{{ specification.user }}"
+    group: "{{ specification.group }}"
     # Was 0640
     mode: 0644
```
Comment (rafzei) [inline ansible/playbooks/roles/kafka/tasks/setup-kafka.yml]:
```suggestion
    mode: u=rw,go=r
```

File: ansible/playbooks/roles/kafka/tasks/generate-certificates.yml [hunk 11]
```diff
@@ -69,71 +69,71 @@
...
+         -import -file "{{ specification.security.ssl.server.keystore_location | dirname }}/ca-cert" \
+         -storepass {{ specification.security.ssl.server.passwords.keystore }} \
+         -keypass {{ specification.security.ssl.server.passwords.key }}
   when:
     - not trustore_exists.stat.exists
 
 - name: Check if CA certificate is already imported
-  shell: keytool -list -v -keystore {{ specification.kafka_var.security.ssl.server.keystore_location }} \
-         -storepass {{ specification.kafka_var.security.ssl.server.passwords.keystore }} \
+  shell: keytool -list -v -keystore {{ specification.security.ssl.server.keystore_location }} \
```
Comment (rafzei) [inline ansible/playbooks/roles/kafka/tasks/generate-certificates.yml]: I would add `set -o pipefail` here

File: ansible/playbooks/roles/zookeeper/tasks/main.yml [hunk 12]
```diff
@@ -112,7 +70,11 @@
   notify: Restart zookeeper
 
 - name: Link /opt/zookeeper to the right version
-  file: path=/opt/zookeeper state=link src="{{ zookeeper_install_dir }}"
+  file:
+    path: /opt/zookeeper
+    state: link
+    src: "{{ zookeeper_install_dir }}"
+  # force: ?
```
Comment (rafzei) [inline ansible/playbooks/roles/zookeeper/tasks/main.yml]: :smiley_cat:

**Review by rafzei — CHANGES_REQUESTED** [review]

File: ansible/playbooks/roles/kafka/tasks/generate-certificates.yml (hunk 4 above)
Comment (cicharka) [reply ansible/playbooks/roles/kafka/tasks/generate-certificates.yml]: That was the only place where we are not using any shell operators, so I only changed it in this line.

File: ansible/playbooks/roles/kafka/tasks/generate-certificates.yml (hunk 11 above)
Comment (cicharka) [reply ansible/playbooks/roles/kafka/tasks/generate-certificates.yml]: done.

File: ansible/playbooks/roles/zookeeper/tasks/main.yml (hunk 12 above)
Comment (cicharka) [reply ansible/playbooks/roles/zookeeper/tasks/main.yml]: yeah :( fixed now!

File: ansible/playbooks/roles/kafka/tasks/common/download_and_unpack_binary.yml [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+    creates: "{{ kafka_install_dir }}"
+    src: "{{ download_directory }}/{{ kafka_bin_filename }}"
+    dest: /opt
+
+- name: Change ownership on Kafka directory
+  file:
+    path: "{{ kafka_install_dir }}"
+    state: directory
+    owner: kafka
+    group: kafka
```
Comment (erzetpe) [inline ansible/playbooks/roles/kafka/tasks/common/download_and_unpack_binary.yml]: This probably should be taken from vars?

File: ansible/playbooks/roles/kafka/tasks/upgrade/install-upgrade.yml [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+  file:
+    dest: /opt/kafka
+    state: link
+    src: "{{ kafka_install_dir}}"
+    force: yes
+
+- name: Remove previous version binaries
+  file:
+    path: /opt/kafka_{{ scala_version }}-{{ old_kafka_version }}
+    state: absent
```
Comment (erzetpe) [inline ansible/playbooks/roles/kafka/tasks/upgrade/install-upgrade.yml:22]: We should confirm that we have something about backup or put some flag to left the people decision if they want to remove old binaries.

**Review by erzetpe — CHANGES_REQUESTED** [review]

File: ansible/playbooks/roles/kafka/tasks/common/download_and_unpack_binary.yml (hunk 13 above)
Comment (cicharka) [reply ansible/playbooks/roles/kafka/tasks/common/download_and_unpack_binary.yml]: Taken from specification as we let user configure that.

File: ansible/playbooks/roles/kafka/tasks/upgrade/install-upgrade.yml (hunk 14 above)
Comment (cicharka) [reply ansible/playbooks/roles/kafka/tasks/upgrade/install-upgrade.yml:22]: Added note in Upgrade docs

**Review by rafzei — DISMISSED** [review]
LGTM

File: ansible/playbooks/roles/kafka/tasks/common/download_and_unpack_binary.yml [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+- name: Download Kafka binaries
+  include_role:
+    name: download
+    tasks_from: download_file
+  vars:
+    file_name: "{{ kafka_bin_filename }}"
+
+- name: Uncompress the Kafka tar
+  unarchive:
+    remote_src: yes
```
Comment (atsikham) [inline ansible/playbooks/roles/kafka/tasks/common/download_and_unpack_binary.yml]: We agreed some time before to use `true/false` values, please change in each affected place.

File: ansible/playbooks/roles/kafka/tasks/common/stop.yml [hunk 16]
```diff
@@ -0,0 +1,5 @@
+---
```
Comment (atsikham) [inline ansible/playbooks/roles/kafka/tasks/common/stop.yml:1]: Is it needed to be stored in a separate file? Same for start.yml.

File: ansible/playbooks/roles/kafka/tasks/generate-certificates.yml [hunk 17]
```diff
@@ -1,55 +1,55 @@
...
+           -noprompt -storepass {{ specification.security.ssl.server.passwords.keystore }} \
+           -keypass {{ specification.security.ssl.server.passwords.key }} \
+           -dname "CN={{ inventory_hostname }}" -ext SAN="DNS:{{ inventory_hostname }}"
   when:
     - not keystore_exists.stat.exists
 
 - name: Check if signing certificate exists
   stat:
-    path: "{{ specification.kafka_var.security.ssl.server.keystore_location | dirname }}/ca-cert"
+    path: "{{ specification.security.ssl.server.keystore_location | dirname }}/ca-cert"
```
Comment (atsikham) [inline ansible/playbooks/roles/kafka/tasks/generate-certificates.yml:29]: Add `get_*` parameters with `false` value where it's possible.
https://docs.ansible.com/ansible/latest/collections/ansible/builtin/stat_module.html

File: ansible/playbooks/roles/kafka/tasks/generate-certificates.yml [hunk 18]
```diff
@@ -1,55 +1,55 @@
...
   register: signing_certificate_exists
   changed_when: false
   when:
     - groups['kafka'][0] == inventory_hostname
 
 - name: Generate signing certificate
-  shell: openssl req -new -x509 -keyout {{ specification.kafka_var.security.ssl.server.keystore_location | dirname }}/ca-key \
-         -out {{ specification.kafka_var.security.ssl.server.keystore_location | dirname }}/ca-cert \
-         -days {{ specification.kafka_var.security.ssl.server.cert_validity }} \
+  shell: openssl req -new -x509 -keyout {{ specification.security.ssl.server.keystore_location | dirname }}/ca-key \
```
Comment (atsikham) [inline ansible/playbooks/roles/kafka/tasks/generate-certificates.yml]: Change `shell` module to `command` where it's possible, like it's done above.

File: ansible/playbooks/roles/kafka/tasks/generate-certificates.yml [hunk 19]
```diff
@@ -1,55 +1,55 @@
...
 - name: Create kafka certificates directory on Epiphany host
   become: false
   file:
-    path: "{{ specification.kafka_var.security.ssl.server.local_cert_download_path }}"
+    path: "{{ specification.security.ssl.server.local_cert_download_path }}"
     state: directory
   delegate_to: localhost
 
 - name: Fetching files
   fetch:
```
Comment (atsikham) [inline ansible/playbooks/roles/kafka/tasks/generate-certificates.yml]: May be changed to `slurp` not to store local files.

File: ansible/playbooks/roles/kafka/tasks/metrics.yml [hunk 20]
```diff
@@ -52,7 +52,7 @@
     src: file_sd_config.yml.j2
     owner: root
     group: root
-    mode: 0644
+    mode: u=rx,go=r
   delegate_to: "{{ item }}"
   notify: restart prometheus
   with_inventory_hostnames:
```
Comment (atsikham) [inline ansible/playbooks/roles/kafka/tasks/metrics.yml]: I suggest to use `loop` to follow the same standard as we have in other places.

File: ansible/playbooks/roles/kafka/tasks/upgrade/main.yml [hunk 21]
```diff
@@ -1,25 +1,16 @@
 ---
-- name: Include defaults from kafka role
-  include_vars:
-    file: roles/kafka/defaults/main.yml
-    name: kafka_defaults
-
 - name: Include pre-flight checks
-  include_tasks: kafka/preflight-check.yml
+  include_tasks: upgrade/preflight-check.yml
```
Comment (atsikham) [inline ansible/playbooks/roles/kafka/tasks/upgrade/main.yml:3]: In `postgresql` role we have the `preflight` directory, what do you think about doing the same here?

File: ansible/playbooks/roles/kafka/tasks/upgrade/main.yml [hunk 22]
```diff
@@ -1,25 +1,16 @@
...
 
 - name: Get installed Kafka version
   shell: >-
     set -o pipefail &&
     /opt/kafka/bin/kafka-server-start.sh --version | grep Commit | grep -oP '^\d+\.\d+\.\d+'
   register: result
 
-- name: Set common facts
+- name: Set old Kafka version fact
```
Comment (atsikham) [inline ansible/playbooks/roles/kafka/tasks/upgrade/main.yml]: How about usage of the fact's name here? `old_kafka_version `

**Review by atsikham — CHANGES_REQUESTED** [review]
This PR also can be the place for this task - https://github.com/epiphany-platform/epiphany/issues/2235. As I see, openjdk 8 is used only for Kafka and Zookeeper.

File: ansible/playbooks/roles/kafka/tasks/common/stop.yml (hunk 16 above)
Comment (cicharka) [reply ansible/playbooks/roles/kafka/tasks/common/stop.yml:1]: start.yml was already in separate file so I only [moved it to common](https://github.com/epiphany-platform/epiphany/pull/2990/files/13f1487d2d1082019b1e57cc51725dd70d3176f9#diff-42a0c302e962e0a498efae3dd5167585a4703e00bc935533284ff0f58408feff), and I did the same for stop.yml just to be consistent with let's call it, 'common tools'.
I know that it is not the biggest improvement...

File: ansible/playbooks/roles/kafka/tasks/common/download_and_unpack_binary.yml (hunk 15 above)
Comment (cicharka) [reply ansible/playbooks/roles/kafka/tasks/common/download_and_unpack_binary.yml]: Ok, I will review other places as well

File: ansible/playbooks/roles/kafka/tasks/generate-certificates.yml (hunk 18 above)
Comment (cicharka) [reply ansible/playbooks/roles/kafka/tasks/generate-certificates.yml]: Already corrected some of them based on @rafzei comments, but you are right, there are still places that can be improved.

File: ansible/playbooks/roles/kafka/tasks/upgrade/main.yml (hunk 21 above)
Comment (cicharka) [reply ansible/playbooks/roles/kafka/tasks/upgrade/main.yml:3]: I checked it now, cool idea -> but Kafka doesn't have any preflight checks file for apply.

But it doesn't mean we cannot adapt that idea in Kafka:
Task `- name: Check if jmx exporter is available`
from `tasks/main.yml` can be extracted into newly created `tasks/prefilight/apply.yml`, and `upgrade/preflight-check.yml` can be moved into `tasks/preflight/upgrade.yml`.

As additional note, I would encourage ourselves to note that approach and let other developers know about that structure.

**Review by atsikham — DISMISSED** [review]

**Comment by rafzei** [comment]
@cicharka One more thing, could you please run ansible-lint task in devcontainer on Kafka playbook and fix the issues?

**Review by seriva — DISMISSED** [review]

**Review by erzetpe — DISMISSED** [review]

**Review by atsikham — DISMISSED** [review]

**Review by atsikham — DISMISSED** [review]

**Comment by przemyslavic** [comment]
:x: It fails when there is no monitoring component enabled:
```
2022-03-23T11:45:38.0521213Z [38;21m11:45:38 INFO cli.src.ansible.AnsibleCommand - TASK [kafka : delegated | create prometheus system group] **********************[0m
2022-03-23T11:45:38.0781789Z [31;21m11:45:38 ERROR cli.src.ansible.AnsibleCommand - fatal: [ci-kfkawsrhelflannel-kafka-vm-0 -> {{ node }}]: FAILED! => {"msg": "'dict object' has no attribute 'prometheus'"}[0m
2022-03-23T11:45:38.1062569Z [31;21m11:45:38 ERROR cli.src.ansible.AnsibleCommand - fatal: [ci-kfkawsrhelflannel-kafka-vm-1 -> {{ node }}]: FAILED! => {"msg": "'dict object' has no attribute 'prometheus'"}[0m
2022-03-23T11:45:38.1160160Z [31;21m11:45:38 ERROR cli.src.ansible.AnsibleCommand - fatal: [ci-kfkawsrhelflannel-kafka-vm-2 -> {{ node }}]: FAILED! => {"msg": "'dict object' has no attribute 'prometheus'"}[0m
2022-03-23T11:45:38.1162720Z [38;21m11:45:38 INFO cli.src.ansible.AnsibleCommand - [0m
```

**Review by seriva — DISMISSED** [review]

**Review by erzetpe — DISMISSED** [review]

**Review by seriva — DISMISSED** [review]

**Review by erzetpe — DISMISSED** [review]

**Review by przemyslavic — APPROVED** [review]

**Review by atsikham — APPROVED** [review]

**Pull request merged**